README
amplayd - Daemon to play movies on an ARCADEmini

Blinkenlights Arcade was a light installation at the
Bibliothèque nationale de France in Paris in 2002, realized
by the Chaos Computer Club. See
http://www.blinkenlights.de/arcade/index.en.html

ARCADEmini is a custom hardware project to create small
replicas of the installation using LEDs. See
http://arcademini.schuermans.info/

The amplayd plays animations on this hardware in standalone
operation. This package is only useful if you have such a
device.

We use it for our installation "BlinkenKuze".
About BlinkenKuze: http://wiki.uplug.de/index.php?page=BlinkenKUZE


debian/changelog
amplayd (0.1.5-1) unstable; urgency=low

  * If the device does not exist, retry instead of giving up. If the device
    does not exist on startup, the daemon waits and tries to open it every few
    seconds. If the device disappears during operation, it should now be
    reopened.
  * Corrects a bunch of errors in the playlist management.

 -- Hermann Schwarting <debian@knackich.de>  Sat, 21 Apr 2007 23:57:29 +0200

amplayd (0.1.4-1) unstable; urgency=low

  * Sane memory management
  * Example synchronization script sync.sh
  * Explain the playlist concept in the man page

 -- Hermann Schwarting <debian@knackich.de>  Sun, 25 Mar 2007 19:56:23 +0200

amplayd (0.1.3-1) unstable; urgency=low

  * Long command line options, new options -d, -s
  * Return value indicates success
  * Chown PID file with -u and change group, too
  * Do not create device if it doesn’t exist
  * Sane behavior with empty playlist
  * Cross-compiling, man page, init script

 -- Hermann Schwarting <debian@knackich.de>  Thu, 22 Mar 2007 20:49:35 +0100

amplayd (0.1.2-1) unstable; urgency=low

  * Initial release

 -- Hermann Schwarting <debian@knackich.de>  Thu,  1 Mar 2007 16:08:58 +0100



src/playlist.c
/*
 *  amplayd -- Daemon to play movies on an ARCADEmini
 *  Copyright (C) 2007  Hermann Schwarting <hfs@gmx.de>
 *
 *  This program is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU General Public License
 *  as published by the Free Software Foundation; either version 2
 *  of the License, or (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the
 *  Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
 *  Boston, MA  02110-1301, USA
 */

#include <string.h>
#include "playlist.h"

#define PLAYLIST_MOVIE_SUFFIX ".bml"
#define PRIO_QUEUE_PREFIX "_"

void ptr_array_randomize( GPtrArray *a, guint from, guint to ){
	gpointer swap;
	guint i, j;
	if( to >= a->len ){
		to = a->len - 1;
	}
	for( i = from; i <= to; ++i ){
		j = g_random_int_range( from, to+1 );
		swap = g_ptr_array_index( a, j );
		g_ptr_array_index( a, j ) = g_ptr_array_index( a, i );
		g_ptr_array_index( a, i ) = swap;
	}
}

/* Sort function for sorting GPtrArrays of strings */
static gint strcmp2(gconstpointer a, gconstpointer b)
{
	const char *aa = *(char **) a;
	const char *bb = *(char **) b;

	return strcmp(aa, bb);
}

void playlist_reorder( Playlist *p ){
	if( p->random->len > 0 ){
		ptr_array_randomize( p->random, p->randomPos, p->random->len-1 );
	}
	g_ptr_array_sort( p->sorted, strcmp2 );
}

GPtrArray* choose_queue( Playlist *p, const gchar *file ){
	if( g_ascii_isdigit( file[0] )){
		return p->sorted;
	}
	if( g_str_has_prefix( file, PRIO_QUEUE_PREFIX )){
		return p->prio;
	}
	return p->random;
}

void playlist_add( Playlist *p, const gchar *file ){
	GPtrArray *queue = choose_queue( p, file );
	g_ptr_array_add( queue, g_strdup( file ));
}

void playlist_remove( Playlist *p, const gchar *file ){
	guint i;
	GPtrArray *queue = choose_queue( p, file );
	for( i = 0; i < queue->len; ++i ){
		if( strcmp( file, g_ptr_array_index( queue, i )) == 0 ){
			g_free( g_ptr_array_index( queue, i ));
			g_ptr_array_remove_index( queue, i );
		}
	}
}

void string_array_free( GPtrArray *p ){
	guint i;
	for( i = 0; i < p->len; ++i ){
		g_free( g_ptr_array_index( p, i ));
	}
	g_ptr_array_free( p, TRUE );
}

void playlist_update( Playlist *p ){
	const gchar *file;
	gboolean playlist_changed = FALSE;
	GPtrArray *contents = g_ptr_array_new();

	/* Read current directory content */
	g_dir_rewind( p->dir );
	while( file = g_dir_read_name( p->dir ) ){
		if( g_str_has_suffix( file, PLAYLIST_MOVIE_SUFFIX ) == TRUE ) {
			g_ptr_array_add( contents, g_strdup( file ) );
		}
	}
	g_ptr_array_sort( contents, strcmp2 );

	/* Compare directory content with last time */
	if( p->files->len == 0 && contents->len > 0 ){
		string_array_free( p->files );
		p->files = contents;
		guint i;
		for( i = 0; i < contents->len; ++i ){
			playlist_add( p, g_ptr_array_index( p->files, i ));
		}
		playlist_changed = TRUE;
	} else if( contents->len == 0 && p->files->len > 0 ){
		guint i;
		for( i = 0; i < p->files->len; ++i ){
			playlist_remove( p, g_ptr_array_index( p->files, i ));
		}
		string_array_free( p->files );
		p->files = contents;
		playlist_changed = TRUE;
	} else if( contents->len == 0 && p->files->len == 0 ){
	} else {
		/* Put the last known directory content (p->files) next to the
		 * current state (contents) in two alphanumerically sorted
		 * columns. Run both down from top to bottom.
		 */
		guint c = 0;
		guint f = 0;
		int cmp;
		while( c < contents->len || f < p->files->len ) {
			if( c >= contents->len ){
				cmp = 1;
			} else if( f >= p->files->len ){
				cmp = -1;
			} else {
				cmp = strcmp( g_ptr_array_index( contents, c ),
						g_ptr_array_index( p->files, f ));
			}
			if( cmp < 0 ){
				playlist_add( p, g_ptr_array_index( contents, c ));
				++c;
				playlist_changed = TRUE;
			} else if( cmp > 0 ){
				playlist_remove( p, g_ptr_array_index( p->files, f ));
				++f;
				playlist_changed = TRUE;
			} else {
				++c;
				++f;
			}
		}
		string_array_free( p->files );
		p->files = contents;
	}
	if( playlist_changed ){
		playlist_reorder( p );
	}
}

Playlist* playlist_new( const gchar *path, GError **error ) {
	Playlist *p = g_slice_new( Playlist );

	p->dir = g_dir_open( path, 0, error );
	if( !p->dir ){
		playlist_free( p );
		return NULL;
	}

	p->files = g_ptr_array_new();
	p->random = g_ptr_array_new();
	p->sorted = g_ptr_array_new();
	p->prio = g_ptr_array_new();
	p->randomPos = p->sortedPos = p->prioPos = 0;

	playlist_update( p );

	return p;
}

void playlist_free( Playlist *p ){
	if( p->dir ){
		g_dir_close( p->dir );
	}
	string_array_free( p->files );
	string_array_free( p->random );
	string_array_free( p->sorted );
	string_array_free( p->prio );
	g_slice_free( Playlist, p );
	p = NULL;
}

const gchar* playlist_next( Playlist *p ){
	playlist_update( p );

	/* Play prio queue first */
	if( p->prioPos < p->prio->len ){
		++p->prioPos;
		return g_ptr_array_index( p->prio, p->prioPos-1 );
	}
	/* Empty prio queue when played once */
	if( p->prio->len > 0 && p->prioPos >= p->prio->len ){
		string_array_free( p->prio );
		p->prio = g_ptr_array_new();
		p->prioPos = 0;
	}

	/* Anything to play at all? */
	if( p->prio->len <= 0 && p->sorted->len <= 0 && p->random->len <= 0 ){
		return NULL;
	}

	/* All queues played -> reorder */
	if( p->sortedPos >= p->sorted->len &&
			p->randomPos >= p->random->len ){
		p->randomPos = 0;
		p->sortedPos = 0;
		playlist_reorder( p );
	}

	/*  Play sorted and random queues fairly */
	if( p->sortedPos < p->sorted->len && ( p->random->len <= 0 ||
			(p->sortedPos+1) * p->random->len <
			(p->randomPos+1) * p->sorted->len )
	){
		++p->sortedPos;
		return g_ptr_array_index( p->sorted, p->sortedPos-1 );
	}
	if( p->randomPos < p->random->len ){
		++p->randomPos;
		return g_ptr_array_index( p->random, p->randomPos-1 );
	}
}



src/playlist.h
/*
 *  amplayd -- Daemon to play movies on an ARCADEmini
 *  Copyright (C) 2007  Hermann Schwarting <hfs@gmx.de>
 *
 *  This program is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU General Public License
 *  as published by the Free Software Foundation; either version 2
 *  of the License, or (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the
 *  Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
 *  Boston, MA  02110-1301, USA
 */

#ifndef _PLAYLIST_H_
#define _PLAYLIST_H_


#include <glib.h>
#include "config.h"

/* Playlist
 *
 * A playlist is controlled by the files in one directory. It is divided into
 * three different qeues, in which entries are sorted depending on their
 * filename. All files must end in '.bml'.
 *
 * priority: All files in this queue are played first, but only once. Filenames
 * must begin with an underscore '_'.
 *
 * sorted: All files in this queue are played in alphanumerical order.
 * Filenames must begin with an digit (0--9).
 *
 * random: All other files are played at random.
 *
 * Priority files are served first. After that files from the sorted and random
 * queues are served intermixed and fairly distributed. After all files have
 * been played, the order of the random queue is reshuffled.
 */
typedef struct _Playlist {
	GPtrArray *files;
	GPtrArray *random;
	guint randomPos;
	GPtrArray *sorted;
	guint sortedPos;
	GPtrArray *prio;
	guint prioPos;
	GDir *dir;
} Playlist;

/* playlist_new() creates a new Playlist or returns NULL in case of an error.
 * The playlist should be freed with playlist_free() after use.
 */
Playlist* playlist_new( const gchar* path, GError **error );

/* Returns the next file in the playlist or NULL if the playlist is empty. The
 * directory is checked for changes each time before a filename is returned,
 * and changes take immediate effect.
 */
const gchar* playlist_next();

/* Frees all memory allocated for the playlist and the file names.
 */
void playlist_free( Playlist* p );

#endif


src/amplayd.c
/*
 *  amplayd -- Daemon to play movies on an ARCADEmini
 *  Copyright (C) 2007  Hermann Schwarting <hfs@gmx.de>
 *
 *  This program is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU General Public License
 *  as published by the Free Software Foundation; either version 2
 *  of the License, or (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the
 *  Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
 *  Boston, MA  02110-1301, USA
 */

#include <blib/blib.h>
#include <errno.h>
#include <getopt.h>
#include <libdaemon/daemon.h>
#include <pwd.h>
#include <signal.h>
#include <stdlib.h>
#include <string.h>
#include <sys/types.h>
#include <time.h>
#include <unistd.h>
#include "config.h"
#include "playlist.h"

enum {
	DAEMON_SUCCESS = 1
};

static gboolean RUNNING;

static void usage( const char* name ){
	g_printerr( "%s - Daemon to play movies on an ARCADEmini\n", AMPLAYD_NAME );
	g_printerr( "Usage: %s [OPTIONS]\n", name );
	g_printerr( "\n" );
	g_printerr( "Options:\n" );
	g_printerr( "    -u --user user      Drop privileges to the specified user.\n" );
	g_printerr( "    -d --device device  Device of the ARCADEmini (default: " AMPLAYD_DEVICE ")\n" );
	g_printerr( "    -s --spool-dir dir  Playlist directory (default: " AMPLAYD_PLAYLISTDIR ")\n" );
	g_printerr( "    -f --foreground     Debug mode (run in foreground).\n" );
	g_printerr( "    -h --help           Prints this help information.\n" );
	g_printerr( "    -V --version        Prints the daemons version.\n" );
	g_printerr( "\n" );
}

static void version(){
	g_printerr( "%s version %d.%d.%d\n", AMPLAYD_NAME, VERSION_MAJOR,
			VERSION_MINOR, VERSION_PATCH );
}

static void daemonize( char* name ){
	pid_t pid;

	/* process' name */
	daemon_pid_file_ident = daemon_log_ident = daemon_ident_from_argv0( name );
	daemon_log_use = DAEMON_LOG_AUTO;

	if(( pid = daemon_pid_file_is_running() ) >= 0 ){
		daemon_log( LOG_ERR, "%s is already running as PID %u.\n",
				AMPLAYD_NAME, pid );
		exit( EXIT_FAILURE );
	}

	daemon_retval_init();

	/* fork(), detach from stdin/out/err, set new group, set working dir */
	if(( pid = daemon_fork() ) < 0 ){
		daemon_log( LOG_ERR, "Fork failed.\n" );
		exit( EXIT_FAILURE );
	} else if ( pid ) { // parent
		if( daemon_retval_wait(1) < 0 ){
			g_printerr( "Daemon couldn't start.\n"
					"Please check syslog for error messages.\n" );
			exit( EXIT_FAILURE );
		}
		exit( EXIT_SUCCESS );
	} else { // daemon
		if ( daemon_pid_file_create() < 0 ) {
			daemon_log( LOG_ERR, "Could not create PID file (%s).\n",
					strerror( errno ));
			exit( EXIT_FAILURE );
		}
	}
}

void signal_handler( int sig ){
	switch( sig ){
		case SIGTERM:
		case SIGINT:
		case SIGQUIT:
			RUNNING = FALSE;
	}
}

gboolean play_movie( GIOChannel *output, BMovie *movie, GError **error ){
	gsize size;
	gsize p_size;
	GList *list;
	GIOStatus status;
	struct timespec t;

	BPacket *packet = b_packet_new( movie->width, movie->height,
			movie->channels, movie->maxval, &size );
	g_return_val_if_fail( packet, FALSE );
	p_size = b_packet_size( packet );
	b_packet_hton( packet );

	for( list = movie->frames; list && RUNNING; list = list->next ){
		gsize bytes_written = 0;
		BMovieFrame *frame = list->data;
		memcpy( packet->data, frame->data, size );
		status = g_io_channel_write_chars( output, (guchar *) packet,
				p_size , &bytes_written, error );
		g_return_val_if_fail( status == G_IO_STATUS_NORMAL, FALSE );
		status = g_io_channel_flush( output, error );
		g_return_val_if_fail( status == G_IO_STATUS_NORMAL, FALSE );
		t.tv_sec = frame->duration/1000;
		t.tv_nsec = frame->duration%1000 * 1000000;
		nanosleep( &t, NULL );
	}

	g_free( packet );
	return TRUE;
}

void io_channel_open_wait( GIOChannel** channel, gchar* filename, GError** error ){
	gboolean channel_open_error = FALSE;
	if( *channel != NULL ){
		g_io_channel_shutdown( *channel, TRUE, error );
		g_io_channel_unref( *channel );
		*channel = NULL;
	}
	while( *channel == NULL ){
		*channel = g_io_channel_new_file( filename, "r+", error );
		if( *channel != NULL ){
			break;
		} else if( !( (*error)->code == G_FILE_ERROR_NOENT ||
					(*error)->code == G_FILE_ERROR_NXIO )){
			break;
		}
		/* Display error message only once */
		if( !channel_open_error ){
			daemon_log(LOG_ERR, "Error opening device '%s': %s\n"
					"Retrying...",
					filename, (*error)->message );
			channel_open_error = TRUE;
		}
		g_clear_error( error );
		sleep( 10 );
	}
	if( g_io_channel_set_encoding( *channel, NULL, error )
			!= G_IO_STATUS_NORMAL ){
		daemon_log(LOG_ERR, "Can't enable binary mode: %s\n",
				(*error)->message );
		g_io_channel_shutdown( *channel, TRUE, NULL );
		*channel = NULL;
	}
	return;
}

int main( int argc, char *argv[] ){
	GError *error = NULL;
	BMovie *movie = NULL;
	Playlist *playlist = NULL;
	gchar *device = NULL;
	gchar *spool_dir = NULL;
	const gchar *filename = NULL;
	gchar *filepath = NULL;
	GIOStatus status;
	int o = -1;
	struct passwd *user = NULL;
	gboolean daemonized = TRUE;
	RUNNING = TRUE;

	/* Command line options */
	static const struct option long_options[] = {
		{ "help",       no_argument,       NULL, 'h' },
		{ "version",    no_argument,       NULL, 'V' },
		{ "foreground", no_argument,       NULL, 'f' },
		{ "user",       required_argument, NULL, 'u' },
		{ "device",     required_argument, NULL, 'd' },
		{ "spool-dir",  required_argument, NULL, 's' },
		{ NULL, 0, NULL, 0}
	};

	while (( o = getopt_long( argc, argv, "d:fhs:u:V", long_options, NULL )) >= 0 ){
		switch (o) {
			case 'd':
				device = g_strdup( optarg );
				break;
			case 'f':
				daemonized = FALSE;
				break;
			case 'h':
				usage( argv[0] );
				exit( EXIT_SUCCESS );
				break;
			case 's':
				spool_dir = g_strdup( optarg );
				break;
			case 'u':
				errno = 0;
				if(( user = getpwnam(optarg)) == NULL ){
					g_printerr( "Could not find user '%s' %s\n",
							optarg,
							( errno == 0 ) ? "" : strerror( errno ) );
					exit( EXIT_FAILURE );
				}
				break;
			case 'V':
				version();
				exit( EXIT_SUCCESS );
				break;
			default:
				usage( argv[0] );
				exit( EXIT_FAILURE );
				break;
		}
	}

	if( !device ){
		device = AMPLAYD_DEVICE;
	}
	if( !spool_dir ){
		spool_dir = AMPLAYD_PLAYLISTDIR;
	}
	if( daemonized ){
		daemonize( argv[0] );
	}

	/* drop privileges */
	if( user != NULL ){
		if( chown( daemon_pid_file_proc(), user->pw_uid, user->pw_gid) != 0 ) {
			daemon_log( LOG_ERR,
					"Couldn't chown '%s' to '%s' uid=%d gid=%d\n",
					daemon_pid_file_proc(), user->pw_name,
					user->pw_uid, user->pw_gid );
		}
		if( setgid( user->pw_gid ) != 0 || setuid( user->pw_uid ) != 0 ){
			daemon_log( LOG_ERR, "Could not set uid '%i', gid '%i': %s\n",
					user->pw_uid, user->pw_gid, strerror( errno ));
			exit( EXIT_FAILURE );
		} else {
			daemon_log( LOG_INFO, "Dropping privileges to uid '%i' (%s)\n",
					user->pw_uid, user->pw_name );
		}
	}

	b_init();

	/* Open playlist and output device */
	GIOChannel *am_dev = NULL;
	io_channel_open_wait( &am_dev, device, &error );
	if( !am_dev ){
		daemon_log(LOG_ERR, "Error opening device '%s': %s\n",
				device, error->message );
		g_clear_error( &error );
		exit( EXIT_FAILURE );
	}
	playlist = playlist_new( spool_dir, &error );
	if( !playlist ){
		daemon_log(LOG_ERR, "Error loading playlist directory '%s': %s\n", 
				spool_dir, error->message );
		g_clear_error( &error );
		g_io_channel_shutdown( am_dev, TRUE, NULL );
		exit( EXIT_FAILURE );
	}

	signal( SIGPIPE, SIG_IGN );
	signal( SIGTERM, signal_handler );
	signal( SIGINT, signal_handler );
	signal( SIGQUIT, signal_handler );

	/* Ok, looks good */
	if( daemonized ){
		daemon_retval_send( DAEMON_SUCCESS );
	}

	/* Main Loop */
	gboolean empty_playlist_error = FALSE;
	while( RUNNING ){
		filename = playlist_next( playlist );
		if( !filename ){
			if( !empty_playlist_error ){
				daemon_log( LOG_ERR, "Playlist is empty. Please "
						"check %s\n", spool_dir );
				empty_playlist_error = TRUE;
			}
			sleep( 10 );
			continue;
		}
		else {
			empty_playlist_error = FALSE;
		}
		filepath = g_strconcat( spool_dir, "/", filename, NULL );
		movie = b_movie_new_from_file( filepath, FALSE, &error );
		if( !movie ){
			daemon_log( LOG_ERR, "Error loading '%s': %s\n",
					filepath, error->message );
			g_clear_error( &error );
			continue;
		}

		if( ! play_movie( am_dev, movie, &error )){
			if( error ){
				daemon_log( LOG_ERR, "Error playing '%s': %s\n",
						filepath, error->message );
				g_clear_error( &error );
			}
			/* Try to reopen the device */
			io_channel_open_wait( &am_dev, device, &error );
			if( !am_dev ){
				daemon_log(LOG_ERR, "Error opening device '%s': %s\n",
						device, error->message );
				g_clear_error( &error );
				exit( EXIT_FAILURE );
			}
		}
		g_object_unref( movie );
		g_free( filepath );
	}

	/* Cleanup */
	playlist_free( playlist );
	if( daemonized ){
		if (daemon_pid_file_remove() == -1) {
			daemon_log(LOG_ERR, "Could not remove pid file (%s).\n",
					strerror(errno));
		}
	}
	g_io_channel_shutdown( am_dev, TRUE, NULL );
	g_io_channel_unref( am_dev );
	return EXIT_SUCCESS;
}
